File: ai-reply-back-end/internal/middleware/ratelimit.go
package middleware

import (
	"net/http"
	"sync"
	"time"

	"github.com/aireply/ai-reply-back-end/internal/transport/httpx"
)

// Limiter — жадтағы жылжымалы терезе.
//
// In memory is the right call for a single-instance deployment and the wrong
// call behind more than one replica: swap this type for Redis and nothing else
// changes. Saying that plainly beats shipping something that looks distributed
// and is not.
type Limiter struct {
	mu      sync.Mutex
	hits    map[string][]time.Time
	lastGC  time.Time
	maxKeys int
}

// NewLimiter — лимитер.
func NewLimiter() *Limiter {
	return &Limiter{hits: make(map[string][]time.Time), lastGC: time.Now(), maxKeys: 50000}
}

// Allow — берілген кілт үшін терезеде орын бар ма.
func (l *Limiter) Allow(key string, limit int, window time.Duration) (bool, time.Duration) {
	if limit <= 0 {
		return true, 0
	}
	now := time.Now()
	l.mu.Lock()
	defer l.mu.Unlock()

	if now.Sub(l.lastGC) > 10*time.Minute || len(l.hits) > l.maxKeys {
		for k, v := range l.hits {
			live := filter(v, now, window)
			if len(live) == 0 {
				delete(l.hits, k)
			} else {
				l.hits[k] = live
			}
		}
		l.lastGC = now
	}

	live := filter(l.hits[key], now, window)
	if len(live) >= limit {
		retry := window - now.Sub(live[0])
		l.hits[key] = live
		if retry < time.Second {
			retry = time.Second
		}
		return false, retry
	}
	l.hits[key] = append(live, now)
	return true, 0
}

func filter(values []time.Time, now time.Time, window time.Duration) []time.Time {
	out := values[:0]
	for _, t := range values {
		if now.Sub(t) < window {
			out = append(out, t)
		}
	}
	return out
}

// KeyFunc — лимит кілтін есептеу.
type KeyFunc func(*http.Request) string

// RateLimit — миддлварь түріндегі шектеу.
func RateLimit(l *Limiter, bucket string, limit int, window time.Duration, key KeyFunc) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			ok, retry := l.Allow(bucket+":"+key(r), limit, window)
			if !ok {
				w.Header().Set("Retry-After", itoa(int(retry.Seconds())))
				httpx.Error(w, http.StatusTooManyRequests, httpx.CodeRateLimited,
					"Too many requests. Try again shortly.",
					map[string]any{"retry_after_seconds": int(retry.Seconds())})
				return
			}
			next.ServeHTTP(w, r)
		})
	}
}

func itoa(v int) string {
	if v <= 0 {
		return "1"
	}
	digits := ""
	for v > 0 {
		digits = string(rune('0'+v%10)) + digits
		v /= 10
	}
	return digits
}

File: AI-Reply/backend/src/rateLimit.js
import { config } from './config.js';

/**
 * Per-client sliding-window rate limiting, held in memory.
 *
 * In memory is the right call for a single-process demo and the wrong call for
 * a real deployment behind more than one instance - swap the two maps for Redis
 * and nothing else here changes. Saying that plainly beats shipping something
 * that looks distributed and is not.
 */

const MINUTE_MS = 60 * 1000;
const DAY_MS = 24 * 60 * 60 * 1000;

/** clientId -> array of request timestamps (ms). */
const hits = new Map();

/** Drops clients that have not been seen for a day so the map cannot grow forever. */
function sweep(now) {
  for (const [clientId, timestamps] of hits) {
    const live = timestamps.filter((t) => now - t < DAY_MS);
    if (live.length === 0) hits.delete(clientId);
    else hits.set(clientId, live);
  }
}

let lastSweep = 0;

/**
 * @returns {{allowed: true} | {allowed: false, scope: 'minute'|'day', retryAfterSeconds: number}}
 */
export function consume(clientId) {
  const now = Date.now();

  if (now - lastSweep > 10 * MINUTE_MS) {
    lastSweep = now;
    sweep(now);
  }

  const timestamps = (hits.get(clientId) ?? []).filter((t) => now - t < DAY_MS);

  const inLastMinute = timestamps.filter((t) => now - t < MINUTE_MS);
  if (inLastMinute.length >= config.limits.perMinute) {
    hits.set(clientId, timestamps);
    const oldest = inLastMinute[0];
    return {
      allowed: false,
      scope: 'minute',
      retryAfterSeconds: Math.max(1, Math.ceil((MINUTE_MS - (now - oldest)) / 1000))
    };
  }

  if (timestamps.length >= config.limits.perDay) {
    hits.set(clientId, timestamps);
    const oldest = timestamps[0];
    return {
      allowed: false,
      scope: 'day',
      retryAfterSeconds: Math.max(1, Math.ceil((DAY_MS - (now - oldest)) / 1000))
    };
  }

  timestamps.push(now);
  hits.set(clientId, timestamps);
  return { allowed: true };
}

/** Test seam. */
export function reset() {
  hits.clear();
  lastSweep = 0;
}
